core/src/main/java/org/apache/iceberg/data/GenericRecord.java

```diff
@@ -115,15 +120,27 @@ public Object get(int pos) {
   }
 
   @Override
+  @SuppressWarnings("unchecked")
   public <T> T get(int pos, Class<T> javaClass) {
     Object value = get(pos);
     if (value == null || javaClass.isInstance(value)) {
       return javaClass.cast(value);
+    }
+
+    if (value instanceof LocalDateTime) {
```

Comment:
This is suspicious to me. Generics should not contain values that need to be converted when read. How would a `LocalDateTime`  get into a generic? Those values are supposed to be long in micros.